hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/balancer/ContainerBalancer.java

```diff
@@ -226,78 +256,188 @@ private boolean initializeIteration() {
-   * contains the specified node.
-   *
-   * @param listToSearch List of DatanodeUsageInfo to be searched.
-   * @param node DatanodeUsageInfo to be searched for.
-   * @return true if the specified node is present in listToSearch, otherwise
-   * false.
+   * Match a source datanode with a target datanode and identify the container
+   * to move.
+   * @param potentialTargets List of potential targets to move container to
+   * @return ContainerMoveSelection containing the selected target and container
    */
-  private boolean containsNode(
-      List<DatanodeUsageInfo> listToSearch, DatanodeUsageInfo node) {
-    int index = 0;
-    Comparator<DatanodeUsageInfo> comparator =
-        DatanodeUsageInfo.getMostUsedByRemainingRatio();
-    int size = listToSearch.size();
-    if (size == 0) {
-      return false;
+  private ContainerMoveSelection matchSourceWithTarget(DatanodeDetails source,
+                                    List<DatanodeDetails> potentialTargets) {
+
+    NavigableSet<ContainerID> candidateContainers =
+        selectionCriteria.getCandidateContainers(source);
+
+    if (candidateContainers.isEmpty()) {
+      LOG.info("ContainerBalancer could not find any candidate containers for" +
+          " datanode {}", source.getUuidString());
+      return null;
+    }
+    LOG.info("ContainerBalancer is finding suitable target for source " +
+        "datanode {}", source.getUuidString());
+    ContainerMoveSelection moveSelection =
+        findTargetStrategy.findTargetForContainerMove(
+            source, potentialTargets, candidateContainers,
+            this::canSizeEnterTarget);
+
+    if (moveSelection == null) {
+      LOG.info("ContainerBalancer could not find a suitable target for " +
+          "source node {}.", source.getUuidString());
+      return null;
     }
 
-    if (comparator.compare(listToSearch.get(0),
-        listToSearch.get(size - 1)) < 0) {
-      index =
-          Collections.binarySearch(listToSearch, node, comparator.reversed());
-    } else {
-      index = Collections.binarySearch(listToSearch, node, comparator);
+    LOG.info("ContainerBalancer matched source datanode {} with target " +
+            "datanode {} for container move.", source.getUuidString(),
+        moveSelection.getTargetNode().getUuidString());
+
+    return moveSelection;
+  }
+
+  /**
+   * Checks if limits maxDatanodesToBalance and maxSizeToMove have not been hit.
+   * @return true if conditions pass and balancing can continue, else false
+   */
+  private boolean checkConditionsForBalancing() {
+    if (countDatanodesBalanced + 2 > maxDatanodesToBalance) {
+      LOG.info("Hit max datanodes to balance limit. {} datanodes have already" +
+              " been balanced and the limit is {}.", countDatanodesBalanced,
+          maxDatanodesToBalance);
+      return false;
     }
-    return index >= 0 && listToSearch.get(index).equals(node);
+    if (totalSizeMoved + (long) ozoneConfiguration.getStorageSize(
```

Comment:
suppose totalSizeMoved = 97G  ,maxSizeToMove = 100G  , and we have a closed 2G ( < default 5G) container replica, can we move it?

Reply:
According to this implementation, no. I've done this to avoid searching for a target datanode and container (and fail early) in case the maximum limit is about to be reached. Alternatively, we can perform this check after getting a `ContainerMoveSelection`.

Reply:
Yes, if we check after selecting a target datanode and container. In that case, some extra work would be performed if the check eventually fails.

Reply:
Created a jira for this issue: https://issues.apache.org/jira/browse/HDDS-5526